simulator/infra/macro.h

```diff
@@ -214,4 +214,29 @@ auto test_subtraction_overflow( T_src1 src1, T_src2 src2)
     return std::make_pair( narrow_cast<T>( result), is_overflow);
 }
 
+template<typename T_src1> static inline auto grev_32( T_src1 src1, size_t shamt) {
+    if (shamt &  1) src1 = ((src1 & 0x5555'5555) <<  1) | ((src1 & 0xAAAA'AAAA) >>  1);
+    if (shamt &  2) src1 = ((src1 & 0x3333'3333) <<  2) | ((src1 & 0xCCCC'CCCC) >>  2);
+    if (shamt &  4) src1 = ((src1 & 0x0F0F'0F0F) <<  4) | ((src1 & 0xF0F0'F0F0) >>  4);
+    if (shamt &  8) src1 = ((src1 & 0x00FF'00FF) <<  8) | ((src1 & 0xFF00'FF00) >>  8);
+    if (shamt & 16) src1 = ((src1 & 0x0000'FFFF) << 16) | ((src1 & 0xFFFF'0000) >> 16);
+    return src1;
+}
+
+template<typename T_src1> static inline auto grev_64( T_src1 src1, size_t shamt) {
+    if (shamt &  1) src1 = ((src1 & 0x5555'5555'5555'5555LL) <<  1) |
+                           ((src1 & 0xAAAA'AAAA'AAAA'AAAALL) >>  1);
+    if (shamt &  2) src1 = ((src1 & 0x3333'3333'3333'3333LL) <<  2) |
+                           ((src1 & 0xCCCC'CCCC'CCCC'CCCCLL) >>  2);
+    if (shamt &  4) src1 = ((src1 & 0x0F0F'0F0F'0F0F'0F0FLL) <<  4) |
+                           ((src1 & 0xF0F0'F0F0'F0F0'F0F0LL) >>  4);
+    if (shamt &  8) src1 = ((src1 & 0x00FF'00FF'00FF'00FFLL) <<  8) |
+                           ((src1 & 0xFF00'FF00'FF00'FF00LL) >>  8);
+    if (shamt & 16) src1 = ((src1 & 0x0000'FFFF'0000'FFFFLL) << 16) |
+                           ((src1 & 0xFFFF'0000'FFFF'0000LL) >> 16);
+    if (shamt & 32) src1 = ((src1 & 0x0000'0000'FFFF'FFFFLL) << 32) |
+                           ((src1 & 0xFFFF'FFFF'0000'0000LL) >> 32);
+    return src1;
+}
```

Comment:
I propose to use C++ function overload here.
You need to use the same function name, but different arguments, like:
```c++
static inline auto grev( uint32 value, size_t shamt);
static inline auto grev( uint64 value, size_t shamt);
static inline auto grev( uint128 value, size_t shamt); // May be not implemented, i.e. assert(0);
```

Then, to call it, you won't need a `bitwidth == 32` check, C++ will do it automatically:

```c++
grev( instr->v_src1, shamt);
```

Reply:
Not implemented static inline function does not allow to compile the project. How can I deal with 128 bit registers?

Reply:
```c++
assert(0);
return 0;
```

Reply:
All warnings are treated as errors, so what should be done with shamt and src1?

Reply:
```c++
static inline auto grev( uint128 /* value */, size_t /* shamt */);
```

But there is one more issue, the code will be not covered. I propose to do that:

```c++
static inline auto grev( uint128 /* value */, size_t /* shamt */)
{
    throw( std::runtime_error( "Generalized error is not implemented for RV128"));
}
```

and write a unit test for that using `CHECK_THROW_AS` macro.

Reply:
I don't understand how does `CHECK_THROW_AS` works. Also non-usage of arguments in the 128-bit function causes compilation-time error (warnings).

Reply:
See CATCH2 help: https://github.com/catchorg/Catch2/blob/master/docs/assertions.md#exceptions

Regarding unused arguments - please check my post, it has the solution.

Reply:
i'm stuck with overloading in alu.h

alu.h
```
 template<typename I> static void grev( I* instr) { instr->v_dst = grev( instr->v_src1, shamt_v_src2<typename I::RegisterUInt>( instr)); }
```
macro.h
```
static inline auto grev( uint32 src1, size_t shamt) {
    if (shamt &  1) src1 = ((src1 & 0x5555'5555) <<  1) | ((src1 & 0xAAAA'AAAA) >>  1);
    if (shamt &  2) src1 = ((src1 & 0x3333'3333) <<  2) | ((src1 & 0xCCCC'CCCC) >>  2);
    if (shamt &  4) src1 = ((src1 & 0x0F0F'0F0F) <<  4) | ((src1 & 0xF0F0'F0F0) >>  4);
    if (shamt &  8) src1 = ((src1 & 0x00FF'00FF) <<  8) | ((src1 & 0xFF00'FF00) >>  8);
    if (shamt & 16) src1 = ((src1 & 0x0000'FFFF) << 16) | ((src1 & 0xFFFF'0000) >> 16);
    return src1;
}

static inline auto grev( uint64 src1, size_t shamt) {
    if (shamt &  1) src1 = ((src1 & 0x5555'5555'5555'5555LL) <<  1) |
                           ((src1 & 0xAAAA'AAAA'AAAA'AAAALL) >>  1);
    if (shamt &  2) src1 = ((src1 & 0x3333'3333'3333'3333LL) <<  2) |
                           ((src1 & 0xCCCC'CCCC'CCCC'CCCCLL) >>  2);
    if (shamt &  4) src1 = ((src1 & 0x0F0F'0F0F'0F0F'0F0FLL) <<  4) |
                           ((src1 & 0xF0F0'F0F0'F0F0'F0F0LL) >>  4);
    if (shamt &  8) src1 = ((src1 & 0x00FF'00FF'00FF'00FFLL) <<  8) |
                           ((src1 & 0xFF00'FF00'FF00'FF00LL) >>  8);
    if (shamt & 16) src1 = ((src1 & 0x0000'FFFF'0000'FFFFLL) << 16) |
                           ((src1 & 0xFFFF'0000'FFFF'0000LL) >> 16);
    if (shamt & 32) src1 = ((src1 & 0x0000'0000'FFFF'FFFFLL) << 32) |
                           ((src1 & 0xFFFF'FFFF'0000'0000LL) >> 32);
    return src1;
}

static inline auto grev( uint128 /* src1 */, size_t /* shamt */) {
    throw std::runtime_error( "Generalized reverse is not implemented for RV128");
}
```
 unit_test.cpp
```
TEST_CASE("RISCV RV32 grev")
{
    CHECK( RISCVInstr<uint32>(0x68D6'55B3).get_disasm() == "grev $a1, $a2, $a3");
    std::vector<TestData<uint32>> cases = {
            {0x5555'5555,  0x1, 0xAAAA'AAAA},
            {0x3333'3333,  0x2, 0xCCCC'CCCC},
            {0x1111'1111,  0x3, 0x8888'8888},
            {0x0123'4567,  0x7, 0x80C4'A2E6},
            {0x3333'3333,  0xF, 0xCCCC'CCCC},
            {0x0123'4567, 0x1F, 0xE6A2'C480},
    };
    for (std::size_t i = 0; i < cases.size(); ++i) {
        RISCVInstr<uint32> instr( "grev", 0);
        instr.set_v_src( cases[i].src1, 0);
        instr.set_v_src( cases[i].src2, 1);
        instr.execute();
        INFO( "Iteration: " << i);
        CHECK( instr.get_v_dst() == cases[i].dst);
    }
}


TEST_CASE("RISCV RV64 grev")
{
    std::vector<TestData<uint64>> cases = {
            {0x5555'5555'5555'5555LL,  0x1, 0xAAAA'AAAA'AAAA'AAAALL},
            {0x3333'3333'3333'3333LL,  0x2, 0xCCCC'CCCC'CCCC'CCCCLL},
            {0x1111'1111'1111'1111LL,  0x3, 0x8888'8888'8888'8888LL},
            {0x0123'4567'89AB'CDEFLL,  0x7, 0x80C4'A2E6'91D5'B3F7LL},
            {0x0123'4567'89AB'CDEFLL,  0xF, 0xC480'E6A2'D591'F7B3LL},
            {0x0123'4567'89AB'CDEFLL, 0x1F, 0xE6A2'C480'F7B3'D591LL},
            {0x0123'4567'89AB'CDEFLL, 0x3F, 0xF7B3'D591'E6A2'C480LL},
    };
    for (std::size_t i = 0; i < cases.size(); ++i) {
        RISCVInstr<uint64> instr( "grev", 0);
        instr.set_v_src( cases[i].src1, 0);
        instr.set_v_src( cases[i].src2, 1);
        instr.execute();
        INFO( "Iteration: " << i);
        CHECK( instr.get_v_dst() == cases[i].dst);
    }
}

TEST_CASE("RISCV RV128 grev")
{
    RISCVInstr<uint128> instr( "grev", 0);
    instr.set_v_src( 0, 0);
    instr.set_v_src( 0, 1);
    CHECK_THROW_AS(instr.execute(), std::runtime_error);
}
```

Reply:
Looks good, what is the question?

Reply:
There is an error:  
error: no matching function for call to ‘ALU::grev(unsigned int&, size_t)’
     template<typename I> static void grev( I* instr) { instr->v_dst = grev( instr->v_src1, shamt_v_src2<typename I::RegisterUInt>( instr)); }
                                                                       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~

Reply:
You have naming conflict between ALU::grev and standalone grev. Please rename standalone function to something different and more meaningful.